File: ios/Sources/BlueCortexCapacitorPowersyncSupabasePlugin/PowerSyncManager.swift
import Foundation
import PowerSync

@MainActor
public final class PowerSyncManager {
    public let connector: SupabaseConnector
    public let schema: Schema
    public let db: PowerSyncDatabaseProtocol
    
    private var watchCallbacks: [String: Task<Void, Never>] = [:]
    private var nextWatchId = 0

    public init(powersyncUrl: String, supabaseUrl: String, supabaseAnonKey: String, schemaDefinition: [[String: Any]], dbFilename: String = "powersync.db") throws {
        connector = SupabaseConnector(
            powerSyncUrl: powersyncUrl,
            supabaseUrl: supabaseUrl,
            supabaseAnonKey: supabaseAnonKey
        )
        
        // Build schema from JSON definition
        schema = try Self.buildSchema(from: schemaDefinition)
        
        db = PowerSyncDatabase(
            schema: schema,
            dbFilename: dbFilename
        )
    }
    
    private static func buildSchema(from definition: [[String: Any]]) throws -> Schema {
        var tables: [Table] = []
        
        for tableDict in definition {
            guard let tableName = tableDict["name"] as? String,
                  let columnsArray = tableDict["columns"] as? [[String: Any]] else {
                throw NSError(domain: "PowerSyncManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid table definition"])
            }
            
            var columns: [Column] = []
            for columnDict in columnsArray {
                guard let columnName = columnDict["name"] as? String,
                      let columnType = columnDict["type"] as? String else {
                    throw NSError(domain: "PowerSyncManager", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid column definition"])
                }
                
                let column: Column
                switch columnType.uppercased() {
                case "TEXT":
                    column = Column.text(columnName)
                case "INTEGER":
                    column = Column.integer(columnName)
                case "REAL":
                    column = Column.real(columnName)
                default:
                    throw NSError(domain: "PowerSyncManager", code: 3, userInfo: [NSLocalizedDescriptionKey: "Unknown column type: \(columnType)"])
                }
                columns.append(column)
            }
            
            // Handle indexes if present
            var indexes: [Index] = []
            if let indexesArray = tableDict["indexes"] as? [[String: Any]] {
                for indexDict in indexesArray {
                    guard let indexName = indexDict["name"] as? String,
                          let indexColumns = indexDict["columns"] as? [[String: Any]] else {
                        continue
                    }
                    
                    var indexedColumns: [IndexedColumn] = []
                    for indexColumn in indexColumns {
                        guard let columnName = indexColumn["name"] as? String else {
                            continue
                        }
                        let ascending = indexColumn["ascending"] as? Bool ?? true
                        indexedColumns.append(ascending ? IndexedColumn.ascending(columnName) : IndexedColumn.descending(columnName))
                    }
                    
                    indexes.append(Index(name: indexName, columns: indexedColumns))
                }
            }
            
            let table = Table(name: tableName, columns: columns, indexes: indexes)
            tables.append(table)
        }
        
        return Schema(tables: tables)
    }

    public func setToken(_ token: String) {
        connector.setToken(token)
    }

    public func connect() async throws {
        try await db.connect(
            connector: connector,
            options: ConnectOptions(
                clientConfiguration: SyncClientConfiguration(
                    requestLogger: SyncRequestLoggerConfiguration(
                        requestLevel: .headers
                    ) { message in
                        self.db.logger.debug(message, tag: "SyncRequest")
                    }
                )
            )
        )
    }

    public func disconnect(clearDatabase: Bool = false) async throws {
        if clearDatabase {
            try await db.disconnectAndClear()
        } else {
            try await db.disconnect()
        }
    }

    public func getVersion() async -> String {
        do {
            return try await db.getPowerSyncVersion()
        } catch {
            return error.localizedDescription
        }
    }

    public func execute(sql: String, parameters: [Any] = []) async throws -> [[String: Any]] {
        let _ = try await db.execute(sql: sql, parameters: parameters)
        return []
    }

    public func getAll(sql: String, parameters: [Any] = []) async throws -> [[String: Any]] {
        let results = try await db.getAll(
            sql: sql,
            parameters: parameters,
            mapper: { cursor in
                var row: [String: Any] = [:]
                for columnName in cursor.columnNames.keys {
                    if let value = try? cursor.getStringOptional(name: columnName) {
                        row[columnName] = value
                    } else if let value = try? cursor.getIntOptional(name: columnName) {
                        row[columnName] = value
                    } else if let value = try? cursor.getDoubleOptional(name: columnName) {
                        row[columnName] = value
                    } else if let value = try? cursor.getBooleanOptional(name: columnName) {
                        row[columnName] = value
                    } else {
                        row[columnName] = NSNull()
                    }
                }
                return row
            }
        )
        return results
    }

    public func getOptional(sql: String, parameters: [Any] = []) async throws -> [String: Any]? {
        let results = try await execute(sql: sql, parameters: parameters)
        return results.first
    }

    public func watch(
        sql: String,
        parameters: [Any] = [],
        callback: @escaping ([[String: Any]]) -> Void
    ) async throws -> String {
        let watchId = "watch_\(nextWatchId)"
        nextWatchId += 1

        let task = Task {
            do {
                for try await results in try db.watch(
                    sql: sql,
                    parameters: parameters,
                    mapper: { cursor in
                        var row: [String: Any] = [:]
                        for columnName in cursor.columnNames.keys {
                            // Try different types since we don't know the column type
                            if let value = try? cursor.getStringOptional(name: columnName) {
                                row[columnName] = value
                            } else if let value = try? cursor.getIntOptional(name: columnName) {
                                row[columnName] = value
                            } else if let value = try? cursor.getDoubleOptional(name: columnName) {
                                row[columnName] = value
                            } else if let value = try? cursor.getBooleanOptional(name: columnName) {
                                row[columnName] = value
                            } else {
                                row[columnName] = NSNull()
                            }
                        }
                        return row
                    }
                ) {
                    callback(results)
                }
            } catch {
                print("Error in watch \(watchId): \(error)")
            }
        }

        watchCallbacks[watchId] = task
        return watchId
    }

    public func unwatch(watchId: String) {
        if let task = watchCallbacks[watchId] {
            task.cancel()
            watchCallbacks.removeValue(forKey: watchId)
        }
    }

    public func writeTransaction(sqlStatements: [String], parameters: [[Any]] = []) async throws {
        try await db.writeTransaction { transaction in
            for (index, sql) in sqlStatements.enumerated() {
                let params = index < parameters.count ? parameters[index] : []
                _ = try transaction.execute(sql: sql, parameters: params)
            }
        }
    }

    public func getSyncStatus() -> [String: Any] {
        let status = db.currentStatus
        return [
            "connected": status.connected,
            "lastSyncedAt": status.lastSyncedAt?.ISO8601Format() ?? "",
            "hasSynced": status.hasSynced,
            "uploading": status.uploading,
            "downloading": status.downloading
        ]
    }
}

File: android/src/main/kotlin/com/bluecortex/mobile/PowerSyncManager.kt
package com.bluecortex.mobile

import android.content.Context
import com.powersync.PowerSyncDatabase
import com.powersync.db.SqlCursor
import com.powersync.db.schema.Column
import com.powersync.db.schema.Index
import com.powersync.db.schema.IndexedColumn
import com.powersync.db.schema.Schema
import com.powersync.db.schema.Table
import kotlin.time.ExperimentalTime
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.atomic.AtomicInteger

class PowerSyncManager(
    context: Context,
    powersyncUrl: String,
    supabaseUrl: String,
    supabaseAnonKey: String,
    schemaDefinition: List<Map<String, Any>>,
    dbFilename: String = "powersync.db"
) {
    val connector: SupabaseConnector
    val schema: Schema
    val db: PowerSyncDatabase

    private val watchJobs = mutableMapOf<String, Job>()
    private val nextWatchId = AtomicInteger(0)
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    init {
        connector = SupabaseConnector(
            powerSyncEndpoint = powersyncUrl,
            supabaseUrl = supabaseUrl,
            supabaseKey = supabaseAnonKey
        )

        schema = buildSchema(schemaDefinition)

        db = PowerSyncDatabase(
            factory = com.powersync.DatabaseDriverFactory(context),
            schema = schema,
            dbFilename = dbFilename
        )
    }

    private fun buildSchema(definition: List<Map<String, Any>>): Schema {
        val tables = definition.map { tableDict ->
            val tableName = tableDict["name"] as? String
                ?: throw IllegalArgumentException("Invalid table definition: missing name")

            @Suppress("UNCHECKED_CAST")
            val columnsData = tableDict["columns"] as? List<Map<String, String>>
                ?: throw IllegalArgumentException("Invalid table definition: missing columns")

            val columns = columnsData.map { columnDict ->
                val columnName = columnDict["name"]
                    ?: throw IllegalArgumentException("Invalid column definition: missing name")
                val columnType = columnDict["type"]
                    ?: throw IllegalArgumentException("Invalid column definition: missing type")

                when (columnType.uppercase()) {
                    "TEXT" -> Column.text(columnName)
                    "INTEGER" -> Column.integer(columnName)
                    "REAL" -> Column.real(columnName)
                    else -> throw IllegalArgumentException("Unknown column type: $columnType")
                }
            }

            // Handle indexes
            @Suppress("UNCHECKED_CAST")
            val indexesData = tableDict["indexes"] as? List<Map<String, Any>> ?: emptyList()
            val indexes = indexesData.mapNotNull { indexDict ->
                val indexName = indexDict["name"] as? String ?: return@mapNotNull null

                @Suppress("UNCHECKED_CAST")
                val indexColumnsData = indexDict["columns"] as? List<Map<String, Any>> ?: return@mapNotNull null

                val indexedColumns = indexColumnsData.mapNotNull { ic ->
                    val colName = ic["name"] as? String ?: return@mapNotNull null
                    val ascending = ic["ascending"] as? Boolean ?: true
                    if (ascending) IndexedColumn.ascending(colName) else IndexedColumn.descending(colName)
                }

                Index(name = indexName, columns = indexedColumns)
            }

            Table(name = tableName, columns = columns, indexes = indexes)
        }

        return Schema(tables)
    }

    fun setToken(token: String) {
        connector.setToken(token)
    }

    suspend fun connect() {
        db.connect(connector)
    }

    suspend fun disconnect(clearDatabase: Boolean = false) {
        if (clearDatabase) {
            db.disconnectAndClear()
        } else {
            db.disconnect()
        }
    }

    suspend fun getVersion(): String {
        return try {
            db.getPowerSyncVersion()
        } catch (e: Exception) {
            e.message ?: "unknown error"
        }
    }

    suspend fun execute(sql: String, parameters: List<Any> = emptyList()): List<Map<String, Any?>> {
        db.execute(sql, parameters)
        return emptyList()
    }

    suspend fun getAll(sql: String, parameters: List<Any> = emptyList()): List<Map<String, Any?>> {
        return db.getAll(sql, parameters) { cursor ->
            val row = mutableMapOf<String, Any?>()
            val colNames = cursor.columnNames
            colNames.forEach { (name, _) ->
                try {
                    val s = cursor.getStringOptional(name)
                    if (s != null) {
                        row[name] = s
                        return@forEach
                    }
                } catch (e: Exception) {
                    // ignore
                }

                try {
                    val l = cursor.getLongOptional(name)
                    if (l != null) {
                        row[name] = l
                        return@forEach
                    }
                } catch (e: Exception) {
                    // ignore
                }

                try {
                    val d = cursor.getDoubleOptional(name)
                    if (d != null) {
                        row[name] = d
                        return@forEach
                    }
                } catch (e: Exception) {
                    // ignore
                }

                try {
                    val b = cursor.getBooleanOptional(name)
                    if (b != null) {
                        row[name] = b
                        return@forEach
                    }
                } catch (e: Exception) {
                    // ignore
                }
                
                row[name] = null
            }
            row
        }
    }

    suspend fun getOptional(sql: String, parameters: List<Any> = emptyList()): Map<String, Any?>? {
        return getAll(sql, parameters).firstOrNull()
    }

    fun watch(
        sql: String,
        parameters: List<Any> = emptyList(),
        callback: (List<Map<String, Any?>>) -> Unit
    ): String {
        val watchId = "watch_${nextWatchId.getAndIncrement()}"

        val job = scope.launch {
            db.watch(sql, parameters) { cursor ->
                val row = mutableMapOf<String, Any?>()
                val colNames = cursor.columnNames
                colNames.forEach { (name, _) ->
                    try {
                        val s = cursor.getStringOptional(name)
                        if (s != null) {
                            row[name] = s
                            return@forEach
                        }
                    } catch (e: Exception) {
                        // ignore
                    }

                    try {
                        val l = cursor.getLongOptional(name)
                        if (l != null) {
                            row[name] = l
                            return@forEach
                        }
                    } catch (e: Exception) {
                        // ignore
                    }

                    try {
                        val d = cursor.getDoubleOptional(name)
                        if (d != null) {
                            row[name] = d
                            return@forEach
                        }
                    } catch (e: Exception) {
                        // ignore
                    }

                    try {
                        val b = cursor.getBooleanOptional(name)
                        if (b != null) {
                            row[name] = b
                            return@forEach
                        }
                    } catch (e: Exception) {
                        // ignore
                    }

                    row[name] = null
                }
                row
            }.collectLatest { results ->
                callback(results)
            }
        }

        watchJobs[watchId] = job
        return watchId
    }

    fun unwatch(watchId: String) {
        watchJobs[watchId]?.cancel()
        watchJobs.remove(watchId)
    }

    suspend fun writeTransaction(sqlStatements: List<String>, parameters: List<List<Any>> = emptyList()) {
        db.writeTransaction { tx ->
            sqlStatements.forEachIndexed { index, sql ->
                val params = if (index < parameters.size) parameters[index] else emptyList()
                tx.execute(sql, params)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    @OptIn(kotlin.time.ExperimentalTime::class)
    fun getSyncStatus(): Map<String, Any> {
        val status = db.currentStatus
        val isoFormatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        isoFormatter.timeZone = TimeZone.getTimeZone("UTC")

        return mapOf(
            "connected" to status.connected,
            "lastSyncedAt" to (status.lastSyncedAt?.let {
                isoFormatter.format(it)
            } ?: ""),
            "hasSynced" to status.hasSynced,
            "uploading" to status.uploading,
            "downloading" to status.downloading
        ) as Map<String, Any>
    }
}

// Private extensions to support optional getters
private inline fun <T> SqlCursor.getColumnValueOptional(
    name: String,
    getValue: (Int) -> T?,
): T? = columnNames[name]?.let { getValue(it) }

private fun SqlCursor.getBooleanOptional(name: String): Boolean? = getColumnValueOptional(name) { getBoolean(it) }
private fun SqlCursor.getDoubleOptional(name: String): Double? = getColumnValueOptional(name) { getDouble(it) }
private fun SqlCursor.getLongOptional(name: String): Long? = getColumnValueOptional(name) { getLong(it) }
private fun SqlCursor.getStringOptional(name: String): String? = getColumnValueOptional(name) { getString(it) }
